### PR TITLE
[ETH] Stepper pure component

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

This PR implements the pure component for the stepper. 

image

- Simple model for the parameters. Its easier to check in cosmos the parameters (its shown as a JSON in the fixture)
- Creates a component with the 3 steps, and 2 progress bars
- Step 1: Models the Send ETH tx state
- Step 2: Models the Order Creation state
- Step 3: Models the Receival of either the tokens or the refund

# Wireframes
- https://www.figma.com/proto/sUwumNFlcslHzShSZJHVsS/2022_COW-Native-ETH-Flow?page-id=1201%3A382&node-id=1232%3A1702&viewport=-96%2C1616%2C0.19&scaling=min-zoom&starting-point-node-id=1204%3A674
- Designs https://www.figma.com/proto/sUwumNFlcslHzShSZJHVsS/2022_COW-Native-ETH-Flow?page-id=1201%3A382&node-id=1298%3A382&viewport=-96%2C1616%2C0.19&scaling=min-zoom&starting-point-node-id=1298%3A382&show-proto-sidebar=1

## Cosmos cases
Cosmos allows to quickly check the happy path, but also many other not so obvious cases that can happen (i.e. the order expires before its creation).

image

## Not included
- I realised we don't have some global styles applied in the cosmos fixtures. I think we should. For example, to include the right fonts. This PR adds a hack to add the fonts in the fixture, but it should be done globally (follow up PR)
- Fine details and possible additional refactors, or restyling will be done by @fairlighteth 

# To Test

1. Go to Cosmos, open `EthFlowStepper` (`command + P` works best to find it)
     * https://pr1402--cowswap.review.gnosisdev.com/cosmos/?fixtureId=%7B%22path%22%3A%22src%2Fcow-react%2Fmodules%2Fswap%2Fpure%2FEthFlow%2FEthFlowStepper%2Findex.cosmos.tsx%22%7D 
2. Use the steps in the control to check how all the cases are handled

### CODE REVIEW

File: src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  const isCreating = state === SmartOrderStatus.CREATING
+  const isFilled = state === SmartOrderStatus.FILLED
+
+  const isRefunding = !!refundTx && !isRefunded
+  const isCanceling = !!cancelationTx && !isCanceled
+
+  const expiredBeforeCreate = isExpired && (isCreating || isIndexing)
+
+  let label: string, stepStatus: StatusIconState, icon: Icon
+  if (expiredBeforeCreate) {
```
Comment (shoom3301) [inline src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx:30]: function getStepConfig(tokenLabel: string, nativeTokenSymbol: string) {
    if (expiredBeforeCreate) {
        return {
            label: 'Receive ' + tokenLabel,
            status: 'pending',
            icon: 'Flag'
        }
    }

    if (isIndexing) {
        return {
            label: 'Receive ' + tokenLabel,
            status: 'not-started',
            icon: 'Flag'
        }
    }

    if (isFilled) {
        return {
            label: 'Received ' + tokenLabel,
            status: 'success',
            icon: 'Flag'
        }
    }

    ...
}

Comment (shoom3301) [inline src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx:30]:
```
if (isIndexing) {
    return {
        label: 'Receive ' + tokenLabel,
        status: 'not-started',
        icon: 'Flag'
    }
}

if (isFilled) {
    return {
        label: 'Received ' + tokenLabel,
        status: 'success',
        icon: 'Flag'
    }
}

...
```

**Comment by elena-zh** [comment]
Hey @anxolin ,great job!

Just some questions: 
1. AFAIK, onchain transaction should handle 'speed up' action, and while the TX is speeded up, a new hash is created. Will we be able to handle this case?
2. It is up for design, however, the stepper does not look good in a mobile view :)
3. Open+Cancelling state: will the order be immediately cancelled? AFAIK, it might not be cancelled, so until the TX is mined/order is cancelled (Open and Cancelled) I don't think we need to show 'Initiating refund' message. WDYT?
image
image

4. Just Expired: I'd show whether the status or the progress bar in red for this case:
image
5. The same is for 'cancelling' onchain transaction (before showing red 2nd step): when the process is initiated, maybe we should show the line in red or so? 
image
or add a status 
image

Thanks!

File: src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx (hunk 1 above)
Comment (anxolin) [reply src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx:30]: Happy to do something in this lines, I even thought about it when I created this Step, however, the config has the label, status, icon, but also the `details`.

Let me see if I can explain my thinking process here.

The construction of the details, and the label/status/icon follow different logics. 

I could make 2 functions, one returns the first part, and the other returns the detail, but both functions would need to either:
- Receive the original parameters, and derive the flags from them (so both of them would do a similar thing)
- Derive the flags from the parameters, and then pass them to this getConfig function (it becomes too cumbersome, and also, when I need to add new conditions, I need to modify the params, which i see it as adding friction)

So considering this, i opted for just a short function that would do this 2 things (first gets the label/icon/state, and the create the details). The refactored part is the one I moved to a generic `Step` component, so this `Step3` can still be fairly small. 

Let me know if the above makes sense. One thing I can do, is to separate more the logic from the presentation also in Step3. I think is overkill given the size of this component, but I will give it a try anyways

Comment (anxolin) [reply src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx:30]: I just gave it a tried to move some of the logic to a config function, so the presentation is simpler, but it feels that is overkill (and makes it harder to make modifications, imo)

I was doing this function, so the JSX can be a bit dumber
image

Maybe if you feel strong about this, we can see ways to improve it together

**Comment by anxolin** [comment]
> AFAIK, onchain transaction should handle 'speed up' action, and while the TX is speeded up, a new hash is created. Will we be able to handle this case?

Yes, we should be able to handle it, although not sure we will (if it takes little time we will --> we have already some logic for this, if it comes for free or almost free we should do it).
This component is agnostic to this, so once we connect it to the logic, is a matter of updating the TX, and the new one will be represented. 

Maybe @alfetopito has an oppinion.


> 2. It is up for design, however, the stepper does not look good in a mobile view :)

I didn't even look there :) 
Yes, this will look WAAAAAAAY better after @fairlighteth  comes to the party. It's currently out of the scope of this PR.

> 3. Open+Cancelling state: will the order be immediately cancelled? AFAIK, it might not be cancelled, so until the TX is mined/order is cancelled (Open and Cancelled) I don't think we need to show 'Initiating refund' message. WDYT?

I went for an optimistic approach. If it's cancelling, we will see (out of this component) the "CANCELING..." state, so you know is not final. But the stepper will show how you won't get the tokens. 

It's true, if it gets executed, it will change into a success. So we could do something else, this is up for discussion. @fairlighteth  WDYT?

File: src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx (hunk 1 above)
Comment (shoom3301) [reply src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/steps/Step3.tsx:30]:
```
const [label, stepStatus, icon] = (() => {
    if (expiredBeforeCreate) {
      return ['Receive ' + tokenLabel, 'pending', Flag]
    }
    if (isIndexing) {
      return ['Receive ' + tokenLabel, 'not-started', Flag]
    }
    if (isFilled) {
      return ['Received ' + tokenLabel, 'success', Check]
    }
    if (isCanceled || isRefunded) {
      return [nativeTokenSymbol + ' Refunded', 'success', Check]
    }
    if (isIndexed) {
      return ['Receive ' + tokenLabel, 'pending', Flag]
    }

    return ['Receive ' + tokenLabel, 'not-started', Flag]
  })()
```

**Comment by anxolin** [comment]
> 4. Just Expired: I'd show whether the status or the progress bar in red for this case:

I will leave this for the design master :) 
I was going to do it, but I talked to @fairlighteth about this, and this things might change a lot. Anyways, I agree looks better with the red progress.

> 5. The same is for 'cancelling' onchain transaction (before showing red 2nd step): when the process is initiated, maybe we should show the line in red or so?

Same as above

File: src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/StatusIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    bgColor: '#f25757',
+    borderColor: '#f25757',
+    color: 'white',
+  },
+}
+
+const LABEL_COLOR: Record<StatusIconState, string> = {
+  success: STATUS_ICON_STYLES['success'].color,
+  pending: STATUS_ICON_STYLES['pending'].color,
+  'not-started': 'gray',
```
Comment (alfetopito) [inline src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/StatusIcon.tsx:39]: Why not use the value defined on `STATUS_ICON_STYLES` like the others?

**Review by alfetopito — APPROVED** [review]

File: src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/StatusIcon.tsx (hunk 2 above)
Comment (anxolin) [reply src/cow-react/modules/swap/pure/EthFlow/EthFlowStepper/StatusIcon.tsx:39]: cause it was hard to read :) I override some.
@fairlighteth will change colors and styles anyways, so doesn't matter too much

**Comment by anxolin** [comment]
@elena-zh I'm merging, we can revisit this after @fairlighteth visit to the stepper

**Pull request merged**